server/api.js

```diff
@@ -57,12 +57,129 @@ const auth = async (req, res, next) => {
+// router.use(auth);
+const authRout = Router();
+authRout.use(auth);
 
 router.get("/", (_, res) => {
 	res.status(200).json({ message: "WELCOME TO LOVE ME TENDER SITE" });
 });
 
+function generateRandomPassword(length = 12) {
+	const lowerCase = "abcdefghijklmnopqrstuvwxyz";
+	const upperCase = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+	const numbers = "0123456789";
+	const specialChars = "!@#$%^&*()_-+=";
+
+	const allChars = lowerCase + upperCase + numbers + specialChars;
+
+	let password = "";
+	password += lowerCase.charAt(Math.floor(Math.random() * lowerCase.length));
+	password += upperCase.charAt(Math.floor(Math.random() * upperCase.length));
+	password += numbers.charAt(Math.floor(Math.random() * numbers.length));
+	password += specialChars.charAt(
+		Math.floor(Math.random() * specialChars.length)
+	);
+
+	for (let i = 4; i < length; i++) {
+		password += allChars.charAt(Math.floor(Math.random() * allChars.length));
+	}
+
+	return password;
+}
+
+function validateEmail(email) {
+	const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
+	return emailRegex.test(email);
+}
+
+router.post("/signup", async (req, res) => {
+	const {
+		email,
+		userType,
+		firstName,
+		lastName,
+		company,
+		address,
+		description,
+	} = req.body;
+
+	const errors = [];
+
+	if (!validateEmail(email)) {
+		errors.push("Invalid email format");
+	}
+
+	if (!["bidder", "contractor"].includes(userType)) {
+		errors.push(
+			"Invalid user type. Allowed values are 'bidder' and 'contractor'"
+		);
+	}
+
+	if (!firstName) {
+		errors.push("First name is required");
+	}
+
+	if (!lastName) {
+		errors.push("Last name is required");
+	}
+
+	if (userType === "contractor" && (!company || !description || !address)) {
+		errors.push(
+			"Company, description, and address are required for contractors"
+		);
+	}
+
+	if (errors.length > 0) {
+		return res.status(400).json({
+			message: "Validation errors",
+			errors: errors,
+		});
+	}
+
+	const client = await pool.connect();
+
+	try {
+		await client.query("BEGIN");
+
+		const password = generateRandomPassword();
+		const hashedPassword = await bcrypt.hash(password, 10);
+
+		const userQuery =
+			"INSERT INTO users (email, password_hash, user_type) VALUES ($1, $2, $3) RETURNING id";
+		const userValues = [email, hashedPassword, userType];
+
+		const userResult = await client.query(userQuery, userValues);
+		const userId = userResult.rows[0].id;
+
+		let userTableQuery;
+		let userTableValues;
+
+		if (userType === "bidder") {
+			userTableQuery =
+				"INSERT INTO bidder (user_id, first_name, last_name, last_update) VALUES ($1, $2, $3, NOW())";
+			userTableValues = [userId, firstName, lastName];
+		} else if (userType === "contractor") {
+			userTableQuery =
+				"INSERT INTO contractor (user_id, company, description, address, last_update) VALUES ($1, $2, $3, $4, NOW())";
+			userTableValues = [userId, company, description, address];
+		} else {
+			await client.query("ROLLBACK");
+			return res.status(400).json({});
```

Comment:
We already validated that the type has to be `bidder` or `contractor` if we get here, it's an error on our side (`500`)